### PR TITLE
fix: updated query in algolia reindex task and added extensive logs.

### PR DESCRIPTION
**Pull request by HammadAhmadWaqas**
## Description

updated query in algolia reindex task and added extensive logs

## Ticket Link

Link to the associated ticket
[5149](https://openedx.atlassian.net/browse/ENT-5149)

## Post-review

Squash commits into discrete sets of changes

### CODE REVIEW

File: enterprise_catalog/apps/api/tasks.py [hunk 1]
```diff
@@ -462,24 +462,21 @@ def index_content_keys_in_algolia(content_keys, algolia_client):
...
     products = []
+    batch_num = 1
+    catalog_uuids_by_key = defaultdict(set)
+    customer_uuids_by_key = defaultdict(set)
+    catalog_queries_by_key = defaultdict(set)
     for content_keys_batch in batch(content_keys, batch_size=TASK_BATCH_SIZE):
-        catalog_uuids_by_key = defaultdict(set)
-        customer_uuids_by_key = defaultdict(set)
-        catalog_queries_by_key = defaultdict(set)
-
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/api/tasks.py:472]: moved them out of the batch loop to share them across all the batches. there will be almost 6000 keys stored in these.

File: enterprise_catalog/apps/api/tasks.py [hunk 2]
```diff
@@ -462,24 +462,21 @@ def index_content_keys_in_algolia(content_keys, algolia_client):
...
         # content_key or parent_content_key or course associated programs. returns courses, programs and course runs.
         query = (
             Q(content_key__in=content_keys_batch)
             | Q(parent_content_key__in=content_keys_batch)
-            | Q(
-                associated_content_metadata__content_key__in=content_keys_batch,
-                associated_content_metadata__id=F('id'),
-                content_type=PROGRAM
-            )
+            | Q(associated_content_metadata__content_key__in=content_keys_batch, content_type=PROGRAM)
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/api/tasks.py:479]: found problem in this query as well. it was not adding course-associated programs in content_metadata queryset.

**Review by saleem-latif — APPROVED** [review]

**Review by sameenfatima78 — APPROVED** [review]

File: enterprise_catalog/apps/api/tasks.py [hunk 3]
```diff
@@ -571,6 +571,25 @@ def index_content_keys_in_algolia(content_keys, algolia_client):
...
+    duplicate_products_count = len(products) - len(log_products.keys())
+    logger.info(
+        f'[ENTERPRISE_CATALOG_ALGOLIA_REINDEX] Found {len(products)} products having {duplicate_products_count}'
+        f'duplicates. Products: {json.dumps(log_products)}'
+    )
+    # -------------------------------->>>>>>>>>>>Logging code ends<<<<<<<<<---------------------
+
+    # There can be possible duplicate products due to course associated programs coming in different batches.
+    # Remove duplicate products having same objectId by keeping the latest one only.
+    products_by_keys = {product["objectID"]: product for product in products}
```
Comment (irfanuddinahmad) [inline enterprise_catalog/apps/api/tasks.py:591]: The objectid would be unique due to the batch id appended at the end of it

File: enterprise_catalog/apps/api/tasks.py [hunk 4]
```diff
@@ -462,24 +462,21 @@ def index_content_keys_in_algolia(content_keys, algolia_client):
         algolia_client: Instance of an Algolia API client
     """
     logger.info(
-        'There are {} total content keys to include in the Algolia index.'.format(len(content_keys))
+        f'[ENTERPRISE_CATALOG_ALGOLIA_REINDEX] There are {len(content_keys)} total content keys to include in the'
+        f' Algolia index.'
     )
     products = []
+    batch_num = 1
+    catalog_uuids_by_key = defaultdict(set)
```
Comment (irfanuddinahmad) [inline enterprise_catalog/apps/api/tasks.py:470]: The batching was done to manage the uuids ... we dont want to share them across batches!!!

**Review by irfanuddinahmad — CHANGES_REQUESTED** [review]

File: enterprise_catalog/apps/api/tasks.py (hunk 4 above)
Comment (HammadAhmadWaqas) [reply enterprise_catalog/apps/api/tasks.py:470]: I think this batching was done to process 250 content_metadata records at a time. To split uuids in 100 100 split records we have different method

File: enterprise_catalog/apps/api/tasks.py (hunk 3 above)
Comment (HammadAhmadWaqas) [reply enterprise_catalog/apps/api/tasks.py:591]: Not really, duplicate records can be made if a program comes again in a different batch.

**Review by irfanuddinahmad — APPROVED** [review]
LGTM!

**Pull request merged**